docs/how-to/push-update-on-mainnet.md

```diff
@@ -0,0 +1,120 @@
+Overview
+========
+
+Software update is a protocol mechanism that allows nodes to agree on software changes without altering protocol constants and update to this software version.
+1. Software update is proposed. Proposal is a datatype that gets into
+   the blockchain. It contains information about version changes and
+   hashes of update files.
+2. Update files are uploaded to the S3 bucket.
+3. Software update is confirmed by voting from majority of nodes.
+4. Nodes that see an update try to download and apply it
+   automatically.
+
+Now, again, step-by-step.
+
+Prerequisites
+=============
+1. Cluster nodes' keys. They are crucial to have since cluster nodes
+   are the only nodes that have stake in the system so they are the
+   only nodes that can propose an update and vote for it.
+2. Access to the S3 bucket
+3. Installers corresponding to software version we're pushing to cluster
+   - Those are two executables (exe for win, pkg for mac) which are to be provided by QA (after QA procedures passed)
+   - Advice: after providing installers, ask QA to confirm hashes are same, we don't want to ship users incorrect installers to users
+4. Software version
+   - Single integer, denote application version as for installers provided
```

Comment:
How is this version number determined?

Reply:
By the `applicationVersion` key in `configuration.yaml`. The current version is 0, and thus any number bigger than 0 will be acceptable.

Reply:
(But ideally you should specify `applicationVersion` matching the version in `configuration.yaml` contained in the new installer)

Reply:
Upd: I'm wrong, the current version for mainnet is `1`, thus the number should be at least `2`.

Reply:
Should each software version be `1 +` the latest applied update?

Suppose over a short period of time we need to propose multiple updates. Is the order in which they're ratified important? For instance, if version `4` is ratified before `3` will the update system apply `4` immediately?

Reply:
Upd: I'm wrong again 🤕 Some users are running `0`, some users are running `1`. Thus we'll need two updates (first we'll upgrade everyone to `1`, then we'll upgrade everyone to `2`). I or someone else will provide more details soon.

Reply:
Ok, I'm told that devops should know which commit corresponds to 1, so we'll need to take an installer for v1, propose it as an update, then take an installer for v2 and propose it as well.

Reply:
@domenkozar, are you clear on what rev corresponds to v1?

@neongreen, how long after v1 is ratified (majority of stake votes yes) should devops wait to propose and ratify v2 to be sure no nodes get v2 before v1? Even better, please answer questions 2-5 at the end of the PR.

Reply:
> how long after v1 is ratified (majority of stake votes yes) should devops wait to propose and ratify v2 to be sure no nodes get v2 before v1?

Nodes can't get v2 before v1, as the current “confirmed version” on the blockchain is v0 (despite some nodes having v1) and the proposal with v2 would get rejected. After v1 becomes confirmed, we can push the v2 update. Unless I'm mistaken, the timeline is as follows:

* v1 is pushed and voted on
* `k` blocks later, v1 becomes confirmed and nodes start downloading it
* immediately after v1 becomes confirmed, we can push v2

Reply:
Suppose the latest ratified update is version `5`, and we try to:

1. propose an update tagged as version `7`,
2. have majority vote in favor of `7`, and
3. wait for version `7` to be applied.

What's the first of those 3 events to fail due to version `6` getting skipped, and how will it manifest?

If it doesn't fail on event (1), how should next update be versioned to ensure it is applied (and that `7` isn't because it may not be compatible with `6`)? Would the version after that need to be `7` or `8`?

Reply:
@neongreen, this sub-thread seems to imply that it is impossible to ratify `vN+2`, without ratifying `vN+1` first (having `vN` ratified as a precondition, naturally).  Is this correct?

Reply:
It will fail at step (1) – nodes would reject the proposal and it wouldn't be included into blocks. You can verify that by grepping the first several chars of a proposal ID in papertrail:

> Oct 03 16:32:56 mainnet.ec2.p-b-4 cardano-node-start: [node.server.update:DEBUG:ThreadId 244000] [2017-10-03 13:32:55 UTC] Processing of proposal csl-daedalus:3 { block v0.0.0, UpId: 53c6d13b, { scripts v0, slot duration (mcs): 20000, block size limit: 1.907 MiB, header size limit: 1.907 MiB, tx size limit: 4 KiB, proposal size limit: 700 B, mpc threshold: 20000000000000/1000000000000000 (approx. 0.02), heavyweight delegation threshold: 300000000000/1000000000000000 (approx. 0.0003), update vote threshold: 1000000000000/1000000000000000 (approx. 0.001), update proposal threshold: 100000000000000/1000000000000000 (approx. 0.1), update implicit period (slots): 10000, softfork rule: no change, tx fee policy: no change, unlock stake epoch: no change }, tags: [patak], no attributes } with id 53c6d13b failed: proposal 53c6d13b has wrong software version for app csl-daedalus **(last known is 0, proposal contains 3)**

(Unfortunately, it looks like this message won't be logged into a locally running node. Papertrail it is.)

Reply:
> last known is 0, proposal contains 3

@neongreen, to clarify -- is "known" == "ratified" here?

Reply:
@deepfire it means "adopted". Blockchain uniquely defines last software version.

Reply:
@volhovM there are two things to be documented here:

a) for wallets, version is determined from `applicationVersion` key in configuration.yaml

b) version number is monotonically increasing and the update system can't skip a single digit (1 -> 3)